### PR TITLE
misc: Implement Python module importing from `.irdl` files

### PR DESCRIPTION
**Pull request by PapyChacal**
Stacked on:
- #2749 

This achieves just importing `.irdl` files (MLIR IR files supposed to contain an irdl dialect) transparently as Python modules, automatically translating the irdl dialect to a PyRDL one and initializing a Python module accordingly, and also generating Python type stubs to be able to programmatically use the said Dialect with type-checking :rocket:

There is *no* way to make pyright benefit from this transparency though, because *it will never run Python code* :face_exhaling: So, I also provide `xdsl-stubgen`, which simply go over `.irdl` files in the `xdsl.dialect` module's directory (Read "including in the install destination if that's what's run"), and importing them, just manually triggering all stubs generation. Add this as a prelude to pyright in CI and Makefile, for the rule of least surprise?

Demo included with `xdsl/dialects/cmath.py` replaced by `xdsl/dialects/cmath.irdl` but registered as any dialect!
NB: This renamed the dialect object to Cmath instead of CMath, because, well, automatic capitalization :shrug: Sounds minimal to me - @math-fehr, maybe we should just use capitalized names in irdl operations?

### CODE REVIEW

**Comment by PapyChacal** [comment]
Note: it worked when running tests sequentially; but failed in CI because Pyright is not triggering the actual cmath import, itself triggering the type stubs generation.

Does anyone have an idea on where to best hook that in?

File: xdsl/ir/context.py
Comment (superlopuh) [inline xdsl/ir/context.py:9]: relatively minor, but if we're splitting things, I'd rather move this to its own file, and it feels a bit more sense to put it in the dialects folder, but that may be a conversation for a different PR

Comment (PapyChacal) [reply xdsl/ir/context.py:9]: Why not, I do think this PR could be split, it's a bit drowning the actual added functionality.
I'll split it after the bases are merged for sanity though :slightly_smiling_face:

File: .gitignore
Comment (math-fehr) [inline .gitignore:159]:
```suggestion
# Type stubs
# These are generated type stubs, we do not otherwise use type stubs in the project
```

File: xdsl/dialects/cmath.py
Comment (math-fehr) [inline xdsl/dialects/cmath.py:1]: Nice

File: xdsl/ir/context.py
Comment (math-fehr) [inline xdsl/ir/context.py:9]: Why do you need the "" here? Can you `from __future__ import annotations`? I forgot if this is supposed to work or not.

Comment (math-fehr) [reply xdsl/ir/context.py:9]: Yeah, I would agree moving it to its own file, for me it's just because I really don't think it goes well with `MLContext`, which is supposedly a dialect-agnostic thing

File: xdsl/__init__.py
Comment (math-fehr) [inline xdsl/__init__.py:18]: Can you add a bit of documentation here and in the next class?

Comment (math-fehr) [reply xdsl/__init__.py:18]: And can we rename the class? :)

File: xdsl/__init__.py
Comment (math-fehr) [inline xdsl/__init__.py:79]: Really nice

**Review by math-fehr — COMMENTED** [review]
Really cool to see this working ;)

File: xdsl/__init__.py
Comment (PapyChacal) [reply xdsl/__init__.py:18]: What a direct attack against state-of-the-art ChatGPT :open_mouth: 
Forgot to rename the boilerplate I kept ahah, will rename more IRDLy

File: xdsl/ir/context.py
Comment (PapyChacal) [reply xdsl/ir/context.py:9]: I'm not sure why annotations? quoted type hints are for not adding runtime dependencies because of type hints. There is a cyclic dependency if I don't use quotes here; Should be solved by moving this function to its own file!

File: .gitignore
Comment (PapyChacal) [reply .gitignore:159]: Good point! But as discussed, we will try to commit the stubs, until we find something more satisfactory for Pyright

Comment (PapyChacal) [reply .gitignore:159]: (more like a self-note)

File: xdsl/ir/context.py
Comment (PapyChacal) [reply xdsl/ir/context.py]: moved `get_all_dialects` to `xdsl.dialects`, and `MLContext` to root `xdsl`, WDYT?

Comment (PapyChacal) [reply xdsl/ir/context.py]: Indeed, solved

**Comment by PapyChacal** [comment]
> would it make sense to put the move from the context object out of the `ir` module in a separate PR? I feel it introduces a lot of noise in this one.

Yes, this is planned as soon as the base is merged

File: .gitignore
Comment (PapyChacal) [reply .gitignore:159]: I'm actually more for the current state I'm offering; aka explicitly generating stub before pyright when we want to be sure. I really don't like committing autogenerated stuff if I can avoid it.
So, adding this clearer comment :yum:

File: xdsl/ir/context.py
Comment (PapyChacal) [reply xdsl/ir/context.py]: All moved out to:
- #2749

File: Makefile
Comment (math-fehr) [inline Makefile:72]:
```suggestion
        # We make sure to generate the python typing stubs before running pyright
	xdsl-stubgen
```

**Review by math-fehr — APPROVED** [review]
Really nice, thanks a lot for splitting the PRs!

File: xdsl/__init__.py
Comment (superlopuh) [inline xdsl/__init__.py:50]: why -5?

Comment (PapyChacal) [reply xdsl/__init__.py:50]: It's "remove the .irdl extension"; would you rather have it be a library call or smth?

Comment (superlopuh) [reply xdsl/__init__.py:50]: I considered suggesting using Path but maybe it's overkill here. It may be worth adding `Drop ".irdl"` as a comment or using Path to remove the extension, your call.

**Review by superlopuh — APPROVED** [review]

File: xdsl/__init__.py
Comment (PapyChacal) [reply xdsl/__init__.py]: Done the call thing!

**Pull request merged**
